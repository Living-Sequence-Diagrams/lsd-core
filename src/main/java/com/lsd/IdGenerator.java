package com.lsd;

import java.util.concurrent.atomic.AtomicInteger;
import java.util.function.Supplier;

import static java.util.UUID.randomUUID;

public class IdGenerator {
    private final AtomicInteger counter = new AtomicInteger();

    private final Supplier<String> idSupplier;

    public IdGenerator(boolean isDeterministic) {
        if (isDeterministic) {
            idSupplier = () -> String.valueOf(counter.incrementAndGet());
        } else {
            idSupplier = () -> randomUUID().toString()
                    .replaceAll("-", "");
        }
    }

    public String next() {
        return idSupplier.get();
    }

    /**
     * Restarts the counter - only works when in deterministic mode
     */
    public void reset() {
        counter.set(0);
    }
}
