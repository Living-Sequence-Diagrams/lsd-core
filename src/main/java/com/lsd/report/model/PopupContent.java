package com.lsd.report.model;

import j2html.tags.ContainerTag;
import j2html.tags.UnescapedText;

import static j2html.TagCreator.*;

/**
 * Simplifies adding content that is only displayed when a hyperlink is clicked - the id used on the href must match.
 */
public class PopupContent {

    public static ContainerTag popupDiv(String id, String title, String content) {
        return div().withId(id).withClass("overlay").attr("onclick", "location.href='#!';").with(
                div().withClass("popup").attr("onclick", "event.stopPropagation();").with(
                        h2(title),
                        a().withClass("close").withHref("#!").with(new UnescapedText("&times;")),
                        div().withClass("content").with(
                                pre(content)
                        )
                )
        );
    }
}
