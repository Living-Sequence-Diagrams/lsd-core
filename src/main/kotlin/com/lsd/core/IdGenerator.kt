package com.lsd.core

import java.util.UUID.randomUUID
import java.util.concurrent.atomic.AtomicInteger

private val randomIdSupplier: () -> String = { randomUUID().toString().replace("-".toRegex(), "") }

class IdGenerator(isDeterministic: Boolean = false) {
    private val counter = AtomicInteger()
    private val deterministicIdSupplier: () -> String = { counter.incrementAndGet().toString() }

    private val idSupplier: () -> String = if (isDeterministic) deterministicIdSupplier else randomIdSupplier

    fun next(): String = idSupplier.invoke()

    /**
     * Restarts the counter - only works when in deterministic mode
     */
    fun reset() = counter.set(0)
}