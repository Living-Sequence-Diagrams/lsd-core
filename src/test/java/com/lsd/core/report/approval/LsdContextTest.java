package com.lsd.core.report.approval;

import com.lsd.core.IdGenerator;
import com.lsd.core.LsdContext;
import com.lsd.core.domain.*;
import com.lsd.core.report.PopupContent;
import org.approvaltests.Approvals;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

import java.util.LinkedHashSet;
import java.util.List;

import static com.lsd.core.builders.ActivateLifelineBuilder.activation;
import static com.lsd.core.builders.DeactivateLifelineBuilder.deactivation;
import static com.lsd.core.builders.MessageBuilder.messageBuilder;
import static com.lsd.core.domain.MessageType.*;
import static com.lsd.core.domain.ParticipantType.*;
import static com.lsd.core.domain.Status.*;
import static j2html.TagCreator.*;

class LsdContextTest {

    private final LsdContext lsdContext = LsdContext.getInstance();

    private final IdGenerator idGenerator = lsdContext.getIdGenerator();

    private final LinkedHashSet<String> additionalIncludes = new LinkedHashSet<>(List.of(
            "tupadr3/font-awesome-5/hamburger",
            "tupadr3/font-awesome-5/heart"
    ));

    private final Participant arnie = ACTOR.called("A", "Arnie");
    private final Participant bettie = BOUNDARY.called("Bettie");
    private final Participant cat = CONTROL.called("Cat");
    private final Participant unused = BOUNDARY.called("Unused participant");

    @BeforeEach
    public void clearContext() {
        lsdContext.clear();
        lsdContext.addParticipants(arnie, bettie, cat, unused);
    }

    @Test
    void createsReportWithScenariosAndEvents() {
        lsdContext.includeFiles(additionalIncludes);

        lsdContext.capture(messageBuilder().id(nextId()).to(arnie).label("in").data("start some job").type(SHORT_INBOUND).build());
        lsdContext.capture(messageBuilder().id(nextId()).from(arnie).to(bettie).label("Message 1").data("some data 1").type(BI_DIRECTIONAL).build());
        lsdContext.capture(messageBuilder().id(nextId()).from(bettie).label("out").data("some data 1").type(SHORT_OUTBOUND).build());
        lsdContext.capture(messageBuilder().id(nextId()).label("An interaction description that is long enough to need abbreviating").from("Beta").to("Gamma").data("β").type(LOST).build());
        lsdContext.capture(messageBuilder().id(nextId()).label("A synchronous response").from("Gamma").to("Beta").data("200 OK").type(SYNCHRONOUS_RESPONSE).build());
        lsdContext.completeScenario("A Success scenario", "Given a first scenario description<br/>When something happens<br/>Then something else happens", SUCCESS);

        lsdContext.capture(messageBuilder().id(nextId()).label("Sending food <$hamburger{scale=0.4}>").from(arnie).to(bettie).colour("orange").build());
        lsdContext.capture(new TimeDelay(null));
        lsdContext.capture("Sending a response from B to A [#red]", "Thank You!");
        lsdContext.capture(new NoteLeft("Friends <$heart{scale=0.4,color=red}>", null));
        lsdContext.addFact("Something to highlight", "Lorem");
        lsdContext.addFact("Something else to highlight", "amet");
        lsdContext.addFact("Something else to highlight", "consectetur");
        lsdContext.addFact("Something else to highlight", "Thank you!");
        lsdContext.completeScenario("A Warning scenario", p(
                text("A popup with a long text that needs scrolling: "),
                a().withHref("#" + "kljasdlfj").withText("click me!"),
                PopupContent.popupDiv("kljasdlfj", "I am popup", ".. Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.")
        ).render(), FAILURE);

        lsdContext.capture("a request from Beta to Gamma", "Please do something");
        lsdContext.capture("sync a synchronous response from Gamma to Beta", "Some Error (123456)");
        lsdContext.addFact("some important value", "123456");
        lsdContext.completeScenario("An Error scenario", "<p>Failure! Expected value to be 123 but was 123456</p>", ERROR);

        Approvals.verify(lsdContext.completeReport("Approval Report").toFile());
    }

    @Test
    void diagramIsSplitOnNewPageEvents() {
        lsdContext.includeFiles(additionalIncludes);

        lsdContext.capture(new PageTitle("A title"));
        lsdContext.capture(messageBuilder().id(nextId()).to(arnie).label("in").data("start some job").type(SHORT_INBOUND).build());
        lsdContext.capture(new Newpage(new PageTitle("A second page")));
        lsdContext.capture(messageBuilder().id(nextId()).from(bettie).label("out").data("some data 1").type(SHORT_OUTBOUND).build());
        lsdContext.capture(new Newpage(new PageTitle("A third page")));
        lsdContext.capture(messageBuilder().id(nextId()).label("An interaction description that is long enough to need abbreviating").from("Beta").to("Gamma").data("β").type(LOST).build());
        lsdContext.completeScenario("A Success scenario", "Given a first scenario description<br/>When something happens<br/>Then something else happens", SUCCESS);

        Approvals.verify(lsdContext.completeReport("Split by NewPage Report").toFile());
    }

    @Test
    void notesCanBeAddedToParticipants() {
        lsdContext.capture(new PageTitle("Adding notes"));
        lsdContext.capture(messageBuilder().id(nextId()).from(arnie).to(bettie).label("Good day to you!").build());
        lsdContext.capture(new NoteLeft("Left note", null));
        lsdContext.capture(new NoteRight("Right note", null));
        lsdContext.capture(new TimeDelay("a few seconds later"));
        lsdContext.capture(messageBuilder().id(nextId()).from(bettie).to(cat).label("Good day to you!").build());
        lsdContext.capture(new NoteLeft("Left of Cat", cat.getComponentName()));
        lsdContext.capture(new NoteRight("Right of Cat", cat.getComponentName()));
        lsdContext.capture(new NoteRight("Right note", null));
        lsdContext.capture(new LogicalDivider("a divider"));
        lsdContext.capture(messageBuilder().id(nextId()).from(cat).to(arnie).label("Me?").build());
        lsdContext.capture(new NoteRight("Right note", null));
        lsdContext.capture(new VerticalSpace(20));
        lsdContext.capture(new NoteLeft("Left of Bettie", bettie.getComponentName()));
        lsdContext.capture(new NoteRight("Right of Bettie", bettie.getComponentName()));

        lsdContext.completeScenario("A Success scenario", "", SUCCESS);

        Approvals.verify(lsdContext.completeReport("Adding notes to participants").toFile());
    }

    @Test
    void participantAreColoured() {
        var greg = ACTOR.called("Greg", null, "green");
        var rosie = BOUNDARY.called("Rosie", null, "red");
        var bart = CONTROL.called("Bart", null, "blue");

        lsdContext.addParticipants(greg, rosie, bart);

        lsdContext.capture(new PageTitle("Adding colours"));
        lsdContext.capture(messageBuilder().id(nextId()).from(greg).to(rosie).label("Roses are red!").build());
        lsdContext.capture(messageBuilder().id(nextId()).from(rosie).to(bart).label("Voilets are blue!").build());
        lsdContext.capture(messageBuilder().id(nextId()).from(bart).to(greg).label("Grass is green?").build());

        lsdContext.completeScenario("A Success scenario", "", SUCCESS);

        Approvals.verify(lsdContext.completeReport("Adding colours to participants").toFile());
    }

    @Test
    void lifelineCanBeActivated() {
        lsdContext.capture(messageBuilder().id(nextId()).from(arnie).to(bettie).label("Good day to you!").build());
        lsdContext.capture(activation().of(bettie).colour("red").build());
        lsdContext.capture(messageBuilder().id(nextId()).from(bettie).to(cat).label("Good day to you!").build());
        lsdContext.capture(deactivation().of(bettie).build());
        lsdContext.capture(activation().of(cat).colour("blue").build());
        lsdContext.capture(messageBuilder().id(nextId()).from(cat).to(arnie).label("Me?").build());
        lsdContext.capture(deactivation().of(cat).build());

        lsdContext.completeScenario("Lifeline activation/deactivation", null, SUCCESS);

        Approvals.verify(lsdContext.completeReport("Activating lifelines").toFile());
    }

    @Test
    void generateCombinedComponentDiagramSource() {
        var dan = CONTROL.called("Dan");

        lsdContext.addParticipants(arnie, bettie, cat, dan);

        lsdContext.capture(messageBuilder().id(nextId()).from(arnie).to(bettie).label("message1").build());
        lsdContext.capture(messageBuilder().id(nextId()).from(bettie).to(cat).label("message2").build());
        lsdContext.capture(messageBuilder().id(nextId()).from(cat).to(arnie).label("message3").build());
        lsdContext.completeScenario("Scenario 1", "", SUCCESS);
        lsdContext.capture(messageBuilder().id(nextId()).from(cat).to(arnie).label("message4").build());
        lsdContext.completeScenario("Scenario 2", "", SUCCESS);
        lsdContext.completeReport("Report 1");

        lsdContext.capture(messageBuilder().id(nextId()).from(bettie).to(arnie).label("message5").build());
        lsdContext.capture(messageBuilder().id(nextId()).from(cat).to(bettie).label("message6").build());
        lsdContext.completeScenario("Scenario 3", "", SUCCESS);
        lsdContext.completeReport("Report 2");

        lsdContext.capture(messageBuilder().id(nextId()).from(bettie).to(dan).label("message5").build());
        lsdContext.capture(messageBuilder().id(nextId()).from(dan).to(arnie).label("message6").build());
        lsdContext.completeScenario("Scenario 4", "", SUCCESS);
        lsdContext.completeReport("Report 3");

        Approvals.verify(lsdContext.completeComponentsReport("Relationships").toFile());
    }

    @Test
    void hasNoDiagramSectionWhenThereAreNoEvents() {
        lsdContext.includeFiles(additionalIncludes);
        lsdContext.addFact("Fact", "no diagrams");
        lsdContext.completeScenario("A scenario without events", "This has no diagrams", SUCCESS);

        Approvals.verify(lsdContext.completeReport("No Diagrams Report").toFile());
    }
    
    @Test
    void participantTypeAndAliasCanBeOverridden() {
        var fred1 = CONTROL.called("Fred");
        var fred2 = DATABASE.called("Fred", "Freddy", "purple");

        lsdContext.addParticipants(List.of(fred1, fred2, bettie));
        lsdContext.capture(messageBuilder().id(nextId()).from("Fred").to(bettie).build());
        lsdContext.completeScenario("Participant is updated", "Freddy is used", SUCCESS);

        Approvals.verify(lsdContext.completeReport("Fred is updated").toFile());
    }

    @Test
    void hasNoDiagramSectionWhenEventsAreCleared() {
        lsdContext.capture(messageBuilder().id(nextId()).to("A").label("in").data("start some job").type(SHORT_INBOUND).build());
        lsdContext.capture(messageBuilder().id(nextId()).to("B").label("in").data("start another job").type(SHORT_INBOUND).build());

        lsdContext.clearScenarioEvents();
        lsdContext.completeScenario("A scenario without events (cleared down)", "This has no diagrams", SUCCESS);

        Approvals.verify(lsdContext.completeReport("No Diagrams Report (cleared down)").toFile());
    }

    @Test
    void createsIndex() {
        lsdContext.completeScenario("Scenario 1", "Error", ERROR);
        lsdContext.completeScenario("Scenario 2", "Warn", FAILURE);
        lsdContext.completeScenario("Scenario 3", "Success", SUCCESS);
        lsdContext.completeReport("First Report (Error)");

        lsdContext.completeScenario("Scenario 1", "Warn", FAILURE);
        lsdContext.completeScenario("Scenario 2", "Success", SUCCESS);
        lsdContext.completeReport("Second Report (Warn)");

        lsdContext.completeScenario("Scenario 1", "Success", SUCCESS);
        lsdContext.completeReport("Third Report");

        lsdContext.completeReport("Fourth Report");

        Approvals.verify(lsdContext.createIndex().toFile());
    }

    @Test
    void generateReport() {
        String result = lsdContext.generateReport("Some title");

        Approvals.verify(result);
    }

    private String nextId() {
        return idGenerator.next();
    }
}