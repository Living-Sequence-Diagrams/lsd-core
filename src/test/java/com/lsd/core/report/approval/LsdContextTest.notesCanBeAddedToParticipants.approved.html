<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" xmlns="http://www.w3.org/1999/html" xml:lang="en" lang="en">
<head>
    <title>Adding notes to participants</title>
    <link rel="stylesheet" href="style.css">
</head>
<body>
<main role="main">
    <h1 class="logo">LSD</h1>
    <h1>Adding notes to participants</h1>
    <article id="4">
        <span onclick="location.href='#top';" title="top of page" class="back_to_top"></span>
        <h2 class="success">A Success scenario</h2>
        
        <section class="sequence diagram svg success">
            <h3 class="success">Sequence Diagram</h3>
            <figure><?xml version="1.0" encoding="us-ascii" standalone="no"?><svg xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" contentStyleType="text/css" height="588px" preserveAspectRatio="none" style="width:489px;height:588px;" version="1.1" viewBox="0 0 489 588" width="489px" zoomAndPan="magnify"><defs/><g><text fill="#000000" font-family="Verdana" font-size="22" font-weight="bold" lengthAdjust="spacing" textLength="163" x="161.5" y="37.1182">Adding notes</text><line style="stroke:#000000;stroke-width:1.0;" x1="94" x2="94" y1="130.752" y2="176.5513"/><line style="stroke:#000000;stroke-width:1.0;stroke-dasharray:1.0,4.0;" x1="94" x2="94" y1="176.5513" y2="217.9199"/><line style="stroke:#000000;stroke-width:1.0;" x1="94" x2="94" y1="217.9199" y2="506.5151"/><line style="stroke:#000000;stroke-width:1.0;" x1="233" x2="233" y1="130.752" y2="176.5513"/><line style="stroke:#000000;stroke-width:1.0;stroke-dasharray:1.0,4.0;" x1="233" x2="233" y1="176.5513" y2="217.9199"/><line style="stroke:#000000;stroke-width:1.0;" x1="233" x2="233" y1="217.9199" y2="506.5151"/><line style="stroke:#000000;stroke-width:1.0;" x1="372.5" x2="372.5" y1="130.752" y2="176.5513"/><line style="stroke:#000000;stroke-width:1.0;stroke-dasharray:1.0,4.0;" x1="372.5" x2="372.5" y1="176.5513" y2="217.9199"/><line style="stroke:#000000;stroke-width:1.0;" x1="372.5" x2="372.5" y1="217.9199" y2="506.5151"/><text fill="#000000" font-family="Verdana" font-size="14" lengthAdjust="spacing" textLength="37" x="73" y="127.8125">Arnie</text><ellipse cx="94.5" cy="61.7373" fill="#FFFFFF" rx="8" ry="8" style="stroke:#000000;stroke-width:1.0;"/><path d="M94.5,69.7373 L94.5,96.7373 M81.5,77.7373 L107.5,77.7373 M94.5,96.7373 L81.5,111.7373 M94.5,96.7373 L107.5,111.7373 " fill="none" style="stroke:#000000;stroke-width:1.0;"/><text fill="#000000" font-family="Verdana" font-size="14" lengthAdjust="spacing" textLength="37" x="73" y="519.5903">Arnie</text><ellipse cx="94.5" cy="531.5298" fill="#FFFFFF" rx="8" ry="8" style="stroke:#000000;stroke-width:1.0;"/><path d="M94.5,539.5298 L94.5,566.5298 M81.5,547.5298 L107.5,547.5298 M94.5,566.5298 L81.5,581.5298 M94.5,566.5298 L107.5,581.5298 " fill="none" style="stroke:#000000;stroke-width:1.0;"/><text fill="#000000" font-family="Verdana" font-size="14" lengthAdjust="spacing" textLength="42" x="209.5" y="127.8125">Bettie</text><path d="M213,85.7373 L213,109.7373 M213,97.7373 L230,97.7373 " fill="none" style="stroke:#000000;stroke-width:1.0;"/><ellipse cx="242" cy="97.7373" fill="#FFFFFF" rx="12" ry="12" style="stroke:#000000;stroke-width:1.0;"/><text fill="#000000" font-family="Verdana" font-size="14" lengthAdjust="spacing" textLength="42" x="209.5" y="519.5903">Bettie</text><path d="M213,526.5298 L213,550.5298 M213,538.5298 L230,538.5298 " fill="none" style="stroke:#000000;stroke-width:1.0;"/><ellipse cx="242" cy="538.5298" fill="#FFFFFF" rx="12" ry="12" style="stroke:#000000;stroke-width:1.0;"/><text fill="#000000" font-family="Verdana" font-size="14" lengthAdjust="spacing" textLength="24" x="357.5" y="127.8125">Cat</text><ellipse cx="372.5" cy="97.7373" fill="#FFFFFF" rx="12" ry="12" style="stroke:#000000;stroke-width:1.0;"/><polygon fill="#000000" points="368.5,85.7373,374.5,80.7373,372.5,85.7373,374.5,90.7373,368.5,85.7373" style="stroke:#000000;stroke-width:1.0;"/><text fill="#000000" font-family="Verdana" font-size="14" lengthAdjust="spacing" textLength="24" x="357.5" y="519.5903">Cat</text><ellipse cx="372.5" cy="538.5298" fill="#FFFFFF" rx="12" ry="12" style="stroke:#000000;stroke-width:1.0;"/><polygon fill="#000000" points="368.5,526.5298,374.5,521.5298,372.5,526.5298,374.5,531.5298,368.5,526.5298" style="stroke:#000000;stroke-width:1.0;"/><polygon fill="#000000" points="221.5,164.5513,231.5,168.5513,221.5,172.5513,225.5,168.5513" style="stroke:#000000;stroke-width:1.0;"/><line style="stroke:#000000;stroke-width:1.0;" x1="94.5" x2="227.5" y1="168.5513" y2="168.5513"/><a href="#1" target="_top" title="Good day to you!" xlink:actuate="onRequest" xlink:href="#1" xlink:show="new" xlink:title="Good day to you!" xlink:type="simple"><text fill="" font-family="Verdana" font-size="13" lengthAdjust="spacing" text-decoration="underline" textLength="115" x="101.5" y="163.8218">Good day to you!</text></a><path d="M10,145.752 L10,170.752 L90,170.752 L90,155.752 L80,145.752 L10,145.752 " fill="#FFFFFF" style="stroke:#000000;stroke-width:1.0;"/><path d="M80,145.752 L80,155.752 L90,155.752 L80,145.752 " fill="#FFFFFF" style="stroke:#000000;stroke-width:1.0;"/><text fill="#000000" font-family="Verdana" font-size="13" lengthAdjust="spacing" textLength="59" x="16" y="163.8218">Left note</text><path d="M238,145.752 L238,170.752 L327,170.752 L327,155.752 L317,145.752 L238,145.752 " fill="#FFFFFF" style="stroke:#000000;stroke-width:1.0;"/><path d="M317,145.752 L317,155.752 L327,155.752 L317,145.752 " fill="#FFFFFF" style="stroke:#000000;stroke-width:1.0;"/><text fill="#000000" font-family="Verdana" font-size="13" lengthAdjust="spacing" textLength="68" x="244" y="163.8218">Right note</text><text fill="#000000" font-family="Verdana" font-size="11" lengthAdjust="spacing" textLength="111" x="178" y="201.6104">a few seconds later</text><polygon fill="#000000" points="360.5,238.7192,370.5,242.7192,360.5,246.7192,364.5,242.7192" style="stroke:#000000;stroke-width:1.0;"/><line style="stroke:#000000;stroke-width:1.0;" x1="233.5" x2="366.5" y1="242.7192" y2="242.7192"/><a href="#2" target="_top" title="Good day to you!" xlink:actuate="onRequest" xlink:href="#2" xlink:show="new" xlink:title="Good day to you!" xlink:type="simple"><text fill="" font-family="Verdana" font-size="13" lengthAdjust="spacing" text-decoration="underline" textLength="115" x="240.5" y="237.9897">Good day to you!</text></a><path d="M377,222.9199 L377,247.9199 L466,247.9199 L466,232.9199 L456,222.9199 L377,222.9199 " fill="#FFFFFF" style="stroke:#000000;stroke-width:1.0;"/><path d="M456,222.9199 L456,232.9199 L466,232.9199 L456,222.9199 " fill="#FFFFFF" style="stroke:#000000;stroke-width:1.0;"/><text fill="#000000" font-family="Verdana" font-size="13" lengthAdjust="spacing" textLength="68" x="383" y="240.9897">Right note</text><path d="M276,258.7192 L276,283.7192 L367,283.7192 L367,268.7192 L357,258.7192 L276,258.7192 " fill="#FFFFFF" style="stroke:#000000;stroke-width:1.0;"/><path d="M357,258.7192 L357,268.7192 L367,268.7192 L357,258.7192 " fill="#FFFFFF" style="stroke:#000000;stroke-width:1.0;"/><text fill="#000000" font-family="Verdana" font-size="13" lengthAdjust="spacing" textLength="70" x="282" y="276.7891">Left of Cat</text><path d="M377,294.5186 L377,319.5186 L477,319.5186 L477,304.5186 L467,294.5186 L377,294.5186 " fill="#FFFFFF" style="stroke:#000000;stroke-width:1.0;"/><path d="M467,294.5186 L467,304.5186 L477,304.5186 L467,294.5186 " fill="#FFFFFF" style="stroke:#000000;stroke-width:1.0;"/><text fill="#000000" font-family="Verdana" font-size="13" lengthAdjust="spacing" textLength="79" x="383" y="312.5884">Right of Cat</text><rect fill="#FFFFFF" height="3" style="stroke:#FFFFFF;stroke-width:1.0;" width="477" x="5" y="346.2175"/><line style="stroke:#000000;stroke-width:0.5;" x1="5" x2="482" y1="346.2175" y2="346.2175"/><line style="stroke:#000000;stroke-width:0.5;" x1="5" x2="482" y1="349.2175" y2="349.2175"/><rect fill="#FFFFFF" height="23.7993" style="stroke:#000000;stroke-width:1.0;" width="80" x="203.5" y="335.3179"/><text fill="#000000" font-family="Verdana" font-size="13" font-weight="bold" lengthAdjust="spacing" textLength="62" x="209.5" y="352.3877">a divider</text><polygon fill="#000000" points="105.5,389.9165,95.5,393.9165,105.5,397.9165,101.5,393.9165" style="stroke:#000000;stroke-width:1.0;"/><line style="stroke:#000000;stroke-width:1.0;" x1="99.5" x2="371.5" y1="393.9165" y2="393.9165"/><a href="#3" target="_top" title="Me?" xlink:actuate="onRequest" xlink:href="#3" xlink:show="new" xlink:title="Me?" xlink:type="simple"><text fill="" font-family="Verdana" font-size="13" lengthAdjust="spacing" text-decoration="underline" textLength="26" x="111.5" y="389.187">Me?</text></a><path d="M377,374.1172 L377,399.1172 L466,399.1172 L466,384.1172 L456,374.1172 L377,374.1172 " fill="#FFFFFF" style="stroke:#000000;stroke-width:1.0;"/><path d="M456,374.1172 L456,384.1172 L466,384.1172 L456,374.1172 " fill="#FFFFFF" style="stroke:#000000;stroke-width:1.0;"/><text fill="#000000" font-family="Verdana" font-size="13" lengthAdjust="spacing" textLength="68" x="383" y="392.187">Right note</text><path d="M120,429.9165 L120,454.9165 L228,454.9165 L228,439.9165 L218,429.9165 L120,429.9165 " fill="#FFFFFF" style="stroke:#000000;stroke-width:1.0;"/><path d="M218,429.9165 L218,439.9165 L228,439.9165 L218,429.9165 " fill="#FFFFFF" style="stroke:#000000;stroke-width:1.0;"/><text fill="#000000" font-family="Verdana" font-size="13" lengthAdjust="spacing" textLength="87" x="126" y="447.9863">Left of Bettie</text><path d="M238,465.7158 L238,490.7158 L355,490.7158 L355,475.7158 L345,465.7158 L238,465.7158 " fill="#FFFFFF" style="stroke:#000000;stroke-width:1.0;"/><path d="M345,465.7158 L345,475.7158 L355,475.7158 L345,465.7158 " fill="#FFFFFF" style="stroke:#000000;stroke-width:1.0;"/><text fill="#000000" font-family="Verdana" font-size="13" lengthAdjust="spacing" textLength="96" x="244" y="483.7856">Right of Bettie</text></g></svg></figure>
            <div onclick="location.href='#!';" id="5" class="overlay">
                <div class="popup" onclick="event.stopPropagation();">
                    <h2>PlantUml markup</h2>
                    <span class="copy" title="copy" onclick="copyToClipboard('sequence_uml_code5')"></span>
                    <a class="close" href="#!">&times;</a>
                    <div class="content">
                        <pre><code id="sequence_uml_code5">@startuml

!theme plain

skinparam backgroundColor transparent

hide unlinked




actor A as &quot;Arnie&quot;

boundary Bettie

control Cat




title Adding notes


A -&gt; Bettie: &lt;text fill&#x3D;&quot;&quot;&gt;[[#1 {Good day to you!} Good day to you!]]&lt;/text&gt;

note left: Left note

note right: Right note

...a few seconds later...

Bettie -&gt; Cat: &lt;text fill&#x3D;&quot;&quot;&gt;[[#2 {Good day to you!} Good day to you!]]&lt;/text&gt;

note left of Cat: Left of Cat

note right of Cat: Right of Cat

note right: Right note

&#x3D;&#x3D; a divider &#x3D;&#x3D;

Cat -&gt; A: &lt;text fill&#x3D;&quot;&quot;&gt;[[#3 {Me?} Me?]]&lt;/text&gt;

note right: Right note

||20||

note left of Bettie: Left of Bettie

note right of Bettie: Right of Bettie


@enduml</code></pre>
                    </div>
                </div>
            </div>
            <sub><a href="#5">[source]</a></sub>
            <sub><a href="#6">[component diagram]</a></sub>
        </section>
        <!-- Popup for component diagram -->
        <div onclick="location.href='#!';" id="6" class="overlay">
            <div class="popup" onclick="event.stopPropagation();">
                <h2>Component Diagram</h2>
                <a class="close" href="#!">&times;</a>
                <div class="content">
                    <pre><?xml version="1.0" encoding="us-ascii" standalone="no"?><svg xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" contentStyleType="text/css" height="270px" preserveAspectRatio="none" style="width:88px;height:270px;" version="1.1" viewBox="0 0 88 270" width="88px" zoomAndPan="magnify"><defs/><g><!--entity A--><g id="elem_A"><ellipse cx="61.5" cy="26" fill="#FFFFFF" rx="8" ry="8" style="stroke:#000000;stroke-width:1.0;"/><path d="M61.5,34 L61.5,61 M48.5,42 L74.5,42 M61.5,61 L48.5,76 M61.5,61 L74.5,76 " fill="none" style="stroke:#000000;stroke-width:1.0;"/><text fill="#000000" font-family="Verdana" font-size="14" lengthAdjust="spacing" textLength="37" x="43" y="92.0752">Arnie</text></g><!--entity Bettie--><g id="elem_Bettie"><path d="M15,135 L15,159 M15,147 L32,147 " fill="none" style="stroke:#000000;stroke-width:1.0;"/><ellipse cx="44" cy="147" fill="#FFFFFF" rx="12" ry="12" style="stroke:#000000;stroke-width:1.0;"/><text fill="#000000" font-family="Verdana" font-size="14" lengthAdjust="spacing" textLength="42" x="14.5" y="177.0752">Bettie</text></g><!--entity Cat--><g id="elem_Cat"><ellipse cx="61.5" cy="232" fill="#FFFFFF" rx="12" ry="12" style="stroke:#000000;stroke-width:1.0;"/><polygon fill="#000000" points="57.5,220,63.5,215,61.5,220,63.5,225,57.5,220" style="stroke:#000000;stroke-width:1.0;"/><text fill="#000000" font-family="Verdana" font-size="14" lengthAdjust="spacing" textLength="24" x="49.5" y="262.0752">Cat</text></g><path d="M51.2987,95.2547 C48.0835,107.3118 44.6285,120.2681 41.7719,130.9805 " fill="none" style="stroke:#000000;stroke-width:1.0;"/><polygon fill="#000000" points="41.7719,130.9805,47.9558,123.3151,43.0602,126.1493,40.2259,121.2538,41.7719,130.9805" style="stroke:#000000;stroke-width:1.0;"/><path d="M42.881,180.0625 C46.3941,191.2773 50.5851,204.6562 54.1005,215.8786 " fill="none" style="stroke:#000000;stroke-width:1.0;"/><polygon fill="#000000" points="54.1005,215.8786,55.2273,206.0944,52.6059,211.1073,47.5931,208.4859,54.1005,215.8786" style="stroke:#000000;stroke-width:1.0;"/><path d="M65.4213,215.8316 C67.0334,204.9512 68.7066,191.8555 69.5,180 C70.9542,158.2708 70.6721,152.7462 69.5,131 C68.8741,119.387 67.6995,106.7933 66.4394,95.3437 " fill="none" style="stroke:#000000;stroke-width:1.0;"/><polygon fill="#000000" points="66.4394,95.3437,63.448,104.7273,66.9864,100.3137,71.4,103.8521,66.4394,95.3437" style="stroke:#000000;stroke-width:1.0;"/></g></svg></pre>
                </div>
                <sub><a href="#uml6">[source]</a></sub>
            </div>
        </div>
        <!-- Popup for component diagram UML via component diagram popup -->
        <div onclick="location.href='#!';" id="uml6" class="overlay">
            <div class="popup" onclick="event.stopPropagation();">
                <h2>PlantUml markup</h2>
                <span class="copy" title="copy" onclick="copyToClipboard('component_uml_code6')"></span>
                <a class="close" href="#6">&times;</a><!--show previous popup on close-->
                <div class="content">
                    <pre><code id="component_uml_code6">@startuml
!pragma layout smetana
!theme plain
skinparam backgroundColor transparent

actor A as &quot;Arnie&quot;
boundary Bettie
control Cat


[A] --&gt; Bettie
[Bettie] --&gt; Cat
[Cat] --&gt; A


remove @unlinked
@enduml</code></pre>
                </div>
            </div>
        </div>
        <!-- Section containing data for sequence diagram popups --><div onclick="location.href='#!';" id="1" class="overlay">
            <div class="popup" onclick="event.stopPropagation();">
                <h2>Good day to you!</h2>
                <a class="close" href="#!">&times;</a>
                <div class="content">
                    <pre></pre>
                </div>
            </div>
        </div><div onclick="location.href='#!';" id="2" class="overlay">
            <div class="popup" onclick="event.stopPropagation();">
                <h2>Good day to you!</h2>
                <a class="close" href="#!">&times;</a>
                <div class="content">
                    <pre></pre>
                </div>
            </div>
        </div><div onclick="location.href='#!';" id="3" class="overlay">
            <div class="popup" onclick="event.stopPropagation();">
                <h2>Me?</h2>
                <a class="close" href="#!">&times;</a>
                <div class="content">
                    <pre></pre>
                </div>
            </div>
        </div>
    </article>
</main>
</body>
<footer>
    <script type="text/javascript" src="https://d3js.org/d3.v4.min.js"></script>
    <link rel="stylesheet" href="https://cdn.jsdelivr.net/gh/highlightjs/cdn-release@11.0.0/build/styles/default.min.css">
    <script src="custom.js"></script>
    <script>
        addSliderForSvgZoom();
        highlightLifelinesWhenClicked();
    </script>
    <script src="https://cdn.jsdelivr.net/gh/highlightjs/cdn-release@11.0.0/build/highlight.min.js"></script>
    <script>hljs.highlightAll();</script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js"></script>
    <script>
        highlightKeywords();
        unMarkSvg();
    </script>
</footer>
</html>
